added sql-data-types branch and corresponding doc (#4700)
influxdata/docs-v2: content/influxdb/cloud-iox/sql-data-types.md

---
title: SQL data types
description: >

>   SQL data types

Reviewer: ```suggestion
  The InfluxDB SQL implementation supports a number of data types including 64-bit integers,
  double-precision floating point numbers, strings, and more.
```

menu:
  influxdb_cloud_iox:
    name: SQL data types
    parent: Query data with SQL
weight: 105
---

> InfluxDB Cloud backed by InfluxDB IOx uses the Apache Arrow DataFusion implementation of SQL. Data types define the type of value that can be stored in table column. In InfluxDB's SQL implementation, a `measurement` is a table, and columns are comprised of `tags`, `fields` and `timestamps`. 

Reviewer: ```suggestion
InfluxDB Cloud backed by InfluxDB IOx uses the [Apache Arrow DataFusion](https://arrow.apache.org/datafusion/) implementation of SQL.
Data types define the type of values that can be stored in table columns.
In InfluxDB's SQL implementation, a **measurement** is structured as a table,
and columns are comprised of **tags**, **fields** and **timestamps**. 
```

> DataFusion uses the Arrow type system for query execution. Data types stored in InfluxDB's storage engine are mapped to SQL data types according to the following tables. 

Reviewer: ```suggestion
DataFusion uses the Arrow type system for query execution.
Data types stored in InfluxDB's storage engine are mapped to SQL data types at query time. 
```

{{% note %}}
When performing casting operations, cast to the **name** of the data type, not the actual data type.  Names are not case sensitive. 

```sql
--Examples
SELECT
'99'::BIGINT
'2019-09-18T00:00:00Z'::timestamp

> ```

Reviewer: ~~~suggestion
When performing casting operations, cast to the **name** of the data type, not the actual data type. 
Names are _case-insensitive_. For example:

```sql
SELECT
  '99'::BIGINT, 
  '2019-09-18T00:00:00Z'::timestamp
```
~~~

{{% /note %}}

## Character types

| Name    | Data type | Description                       |
| :------ | :-------- | --------------------------------- |
| CHAR    | UTF8      | Character string, fixed-length    |
| VARCHAR | UTF8      | Character string, variable-length |
| TEXT    | UTF8      | Variable unlimited length         |

```sql
--Examples
abcdefghijk
time
"h2o_temperature"

> ```

Reviewer: ~~~suggestion
##### Example character types

```sql
abcdefghijk
time
"h2o_temperature"
```
~~~

## Numeric types

> InfluxDB stores all integers as signed 64bit integers. The following numeric types are supported:

Reviewer: InfluxDB supports both signed and unsigned integers.

```suggestion
The following numeric types are supported:
```

| Name            | Data type | Description                  |
| :-------------- | :-------- | :--------------------------- |

> | BIGINT          | INT64     | large-range integer          |
> | BIGINT UNSIGNED | INT64     | large-range unsigned integer |
> | DOUBLE          | FLOAT64   | 64-bit floating-point number |

Reviewer: ```suggestion
| BIGINT          | INT64     | 64-bit signed integer        |
| BIGINT UNSIGNED | INT64     | 64-bit unsigned integer      |
| DOUBLE          | FLOAT64   | 64-bit floating-point number |
```

> Minimum integer:` -9223372036854775808`  
> Maximum integer: `9223372036854775807`

Reviewer: ```suggestion
Minimum signed integer:` -9223372036854775808`  
Maximum signed integer: `9223372036854775807`
```

Minimum unsigned integer (uinteger): `0`  
Maximum unsigned integer (uinteger): `18446744073709551615`  

> Floats can be a decimal point, decimal integer or decimal fraction.

Reviewer: ```suggestion
Floats can be a decimal point, decimal integer, or decimal fraction.
```

```sql
--Examples
23.8
-446.89
5.00
0.033

> ```

Reviewer: ~~~suggestion
##### Example float types

```sql
23.8
-446.89
5.00
0.033
```
~~~

## Time types

A time type is a single point in time using nanosecond precision.  

| Name      | Data type | Description                |
| :-------- | :-------- | :------------------------- |
| TIMESTAMP | TIMESTAMP | TimeUnit::Nanosecond, None |

> 

Reviewer: ```suggestion
```

The following date and time formats are supported:

```sql
--Examples
YYYY-MM-DDT00:00:00.000Z 
YYYY-MM-DDT00:00:00.000-00:00 
YYYY-MM-DD 00:00:00.000-00:00 
YYYY-MM-DDT00:00:00Z
YYYY-MM-DD 00:00:00.000
YYYY-MM-DD 00:00:00

> 1567296000000000000 
> 1566176400 

Reviewer: We need to be careful here. These aren't time types. These are integers. You can cast integers to a TIMESTAMP, but they aren't natively recognized as timestamps.

```suggestion
```

```

## Boolean types

Booleans store TRUE or FALSE values. 

| Name    | Data type | Description                                                           |
| :------ | :-------- | :-------------------------------------------------------------------- |
| BOOLEAN | BOOLEAN   | TRUE or FALSE for strings, 0 and 1 for integers, uintegers and floats |

Booleans are parsed in the following manner:

- string: `TRUE` or `FALSE`

> - integer: 0 is false, 1 is true
> - uinteger: 0 is false, 1 is true
> - float: 0.0 is false, 1.0 is true

Reviewer: ```suggestion
- integer: 0 is false, non-zero is true
- uinteger: 0 is false, non-zero is true
- float: 0.0 is false, non-zero is true
```

> 
> Note that zero vaules are parsed as `false` and non zero negative values are parsed as `true`. 

Reviewer: This is redundant with the change above.

```suggestion
```

```sql
--Examples
true
TRUE
false
FALSE

> ```

Reviewer: ~~~suggestion
##### Example boolean types

```sql
true
TRUE
false
FALSE
```
~~~

> ## Unsupported types
> 
> The following types are not currently supported:
> - UUID
> - BLOB
> - CLOB
> - BINARY
> - VARBINARY
> - REGCLASS
> - NVARCHAR
> - CUSTOM
> - ARRAY
> - ENUM
> - SET
> - INTERVAL
> - DATETIME
> - BYTEA

Reviewer: We should probably include all the non-64-bit numeric types that the InfluxDB doesn't support in this list as well.

```suggestion
## Unsupported SQL types

The following SQL types are not currently supported:

- UUID
- BLOB
- CLOB
- BINARY
- VARBINARY
- REGCLASS
- NVARCHAR
- CUSTOM
- ARRAY
- ENUM
- SET
- INTERVAL
- DATETIME
- BYTEA
```

---
title: SQL data types
description: >
  The InfluxDB SQL implementation supports a number of data types including 64-bit integers,
  double-precision floating point numbers, strings, and more.
menu:
  influxdb_cloud_iox:
    name: SQL data types
    parent: Query data with SQL
weight: 105
---

InfluxDB Cloud backed by InfluxDB IOx uses the [Apache Arrow DataFusion](https://arrow.apache.org/datafusion/) implementation of SQL.
Data types define the type of values that can be stored in table columns.
In InfluxDB's SQL implementation, a **measurement** is structured as a table,
and columns are comprised of **tags**, **fields** and **timestamps**. 

> DataFusion uses the Arrow type system for query execution.

Author: done

| Name      | Data type | Description                |
| :-------- | :-------- | :------------------------- |
| TIMESTAMP | TIMESTAMP | TimeUnit::Nanosecond, None |

The following date and time formats are supported:

```sql
--Examples
YYYY-MM-DDT00:00:00.000Z 
YYYY-MM-DDT00:00:00.000-00:00 
YYYY-MM-DD 00:00:00.000-00:00 
YYYY-MM-DDT00:00:00Z
YYYY-MM-DD 00:00:00.000
YYYY-MM-DD 00:00:00
```

## Boolean types

Booleans store TRUE or FALSE values. 

| Name    | Data type | Description                                                           |
| :------ | :-------- | :-------------------------------------------------------------------- |
| BOOLEAN | BOOLEAN   | TRUE or FALSE for strings, 0 and 1 for integers, uintegers and floats |

Booleans are parsed in the following manner:

- string: `TRUE` or `FALSE`
- integer: 0 is false, non-zero is true
- uinteger: 0 is false, non-zero is true
- float: 0.0 is false, non-zero is true

##### Example boolean types

```sql
true
TRUE
false
FALSE
```

## Unsupported SQL types

The following SQL types are not currently supported:

- UUID
- BLOB
- CLOB
- BINARY
- VARBINARY
- REGCLASS
- NVARCHAR
- CUSTOM
- ARRAY
- ENUM
- SET

> - INTERVAL

Reviewer: @alamb Thanks for calling this out. Do you know the explicit interval literal formats we support? Are they the same as or similar to those listed [here](https://media.datadirect.com/download/docs/openaccess/alloa/index.html#page/sqlref/interval-literals.html)?

> - INTERVAL

Reviewer 2: The logical types supported by arrow (and thus by DataFusion) are listed (somewhat obscurely) in https://docs.rs/arrow/31.0.0/arrow/datatypes/index.html

So in this case the interval types are: 

[IntervalDayTimeType](https://docs.rs/arrow/31.0.0/arrow/datatypes/struct.IntervalDayTimeType.html)	A “calendar” interval type in days and milliseconds.
[IntervalMonthDayNanoType](https://docs.rs/arrow/31.0.0/arrow/datatypes/struct.IntervalMonthDayNanoType.html)	A “calendar” interval type in months, days, and nanoseconds.
[IntervalYearMonthType](https://docs.rs/arrow/31.0.0/arrow/datatypes/struct.IntervalYearMonthType.html)	A “calendar” interval type in months.

That is not quite the same as the SQL equivalent. Here are the types when writing about intervals:

```sql
❯ select arrow_typeof(interval '1 month');
+-------------------------------------+
| arrowtypeof(IntervalYearMonth("1")) |
+-------------------------------------+
| Interval(YearMonth)                 |
+-------------------------------------+
1 row in set. Query took 0.006 seconds.
❯ select arrow_typeof(interval '1 second');
+--------------------------------------+
| arrowtypeof(IntervalDayTime("1000")) |
+--------------------------------------+
| Interval(DayTime)                    |
+--------------------------------------+
1 row in set. Query took 0.000 seconds.
```

> - INTERVAL

Author: added interval and examples
